File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    def get_subset(self,nums:list,ans:list,idx:int,all_subset:list)->None:
        nums.sort()
        if(idx==len(nums)):
            all_subset.append(ans[:])
            return 
        ans.append(nums[idx])
        self.get_subset(nums,ans,idx+1,all_subset)
        ans.pop()
        while(idx+1 < len(nums) and nums[idx]==nums[idx+1]):
            idx += 1
        self.get_subset(nums,ans,idx+1,all_subset)     
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans:list = []
        all_subset:list = []
        self.get_subset(nums,ans,0,all_subset) 
        return all_subset

File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    void subsets(int idx, vector<int>&nums, vector<vector<int>>&result, vector<int>&set){
        if(idx == nums.size()){
            result.push_back({set});
            return;
        }
        set.push_back(nums[idx]);
        subsets(idx+1, nums, result, set);
        set.pop_back();
        while(idx+1<nums.size() && nums[idx]==nums[idx+1])
            idx+=1;
        subsets(idx+1, nums, result, set);    
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>>result;
        vector<int>set;
        subsets(0, nums, result, set);
        return result;
    }
};